lib/posix/pipe.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <posix/unistd.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <sys/fdtable.h>
+
+struct pipe_fd_object {
+	struct stat stat;
+	/* counterpart file descriptor */
+	int other;
+};
+
+/* read side of pipe */
+struct pipe_fd_robject {
```

Comment:
Did you consider a scheme where there's a single `struct`? Given that you alloc it via k_malloc(), would allow to avoid allo overheads. Typechecking may be a problem. (But there's no typechecking so far in your code wither, given that it's mostly placeholder to fill in.)

Reply:
Originally I did want to use a single struct, but then realized I would be duplicating the buffer and pipe objects. It would be possible to use a single struct, but it would require an extra malloc for the buffer and pipe.

Typecheckinf should be doable via the stat. Can check for read / write ends. Also can check use S_ISFIFO() to check the st_mode field (although technically there is a stable check before the callback is made).

Reply:
> Originally I did want to use a single struct, but then realized I would be duplicating the buffer and pipe objects.

I mean single structure for both rd and wr ends. fdtable would refer to the same struct for both rd and we fd's. You then would need to differentiate between the two.

> Typecheckinf should be doable via the stat. 

I mean, internal typechecking. If someone calls write() on the read end, you should not segfault (but return a proper error code or whatever POSIX requires).

> Also can check use S_ISFIFO() to check the st_mode field.

I would say you're thinking in somewhat heavy terms. Stuff like st_mode and S_ISFIFO is external interface. Internally in Zephyr, we should target to save our users' memory (at least, within the kernel). And a single bit is enough to distinguish 2 values internally.

Reply:
> > Originally I did want to use a single struct, but then realized I would be duplicating the buffer and pipe objects.
> 
> I mean single structure for both rd and wr ends. fdtable would refer to the same struct for both rd and we fd's. You then would need to differentiate between the two.

Hmmm... interesting idea. Technically, if it were possible to look up an fd by object pointer, then it would be possible to compare the integer file descriptors. That seems like a lot of work though and it is a hot code path, so it might be best just to 

> > Typecheckinf should be doable via the stat.
> 
> I mean, internal typechecking. If someone calls write() on the read end, you should not segfault (but return a proper error code or whatever POSIX requires).

Yes. Hence, the S_IRUSR and S_IWUSR bits. 

> > Also can check use S_ISFIFO() to check the st_mode field.
> 
> I would say you're thinking in somewhat heavy terms. Stuff like st_mode and S_ISFIFO is external interface. Internally in Zephyr, we should target to save our users' memory (at least, within the kernel). And a single bit is enough to distinguish 2 values internally.

Good point. See my previous comment about it. I'll probably just use a bool to indicate R / W. If Zephyr opts to eventually implement fstat(2), then it can probably also just be done as an ioctl to save space in the vtable.